### PR TITLE
nodeipam: fix hot loop on non-existent node

### LINKED ISSUE #96992: IPAM range allocator hotlooping on non-existent node
**What happened**:
KCM hotloop on
```
range_allocator.go:338] Failed while getting node node1 for updating Node.Spec.PodCIDRs: node "node1" not found
```
Looking in the code it is obvious that this can happen as the error case writes into the channel the select is reading
https://github.com/kubernetes/kubernetes/blob/61dc69ac2cf936ccafb51e9d946097b62b4ead07/pkg/controller/nodeipam/ipam/range_allocator.go#L188-L200

**What you expected to happen**:
range allocator should detect when the Node retrieval fails because the Node doesn't exist and not requeue.  Also requeue should be rate limited.

**How to reproduce it (as minimally and precisely as possible)**:
Seems to be a timing window.  The node has to be created and deleted quickly before the range allocator can process it.

fyi @deads2k @thockin

### PR DESCRIPTION
**Pull request by sjenning**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Fixes a hot loop in the IPAM range allocator.  This was fixed long ago for the cloud CIDR allocator in https://github.com/kubernetes/kubernetes/pull/68084.  This PR just syncs with that implementation.

**Which issue(s) this PR fixes**:
Fixes #96992

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```
/cc @derekwaynecarr @thockin @deads2k

### CODE REVIEW

File: pkg/controller/nodeipam/ipam/range_allocator.go [hunk 1]
```diff
@@ -191,10 +191,21 @@ func (r *rangeAllocator) worker(stopChan <-chan struct{}) {
...
+					klog.V(2).Infof("Retrying update for %q after %v", workItem, timeout)
+					time.AfterFunc(timeout, func() {
+						// Requeue the failed node for update again.
+						r.nodeCIDRUpdateChannel <- workItem
+					})
+					continue
+				}
+				klog.Errorf("Exceeded retry count for %q, dropping from queue", workItem)
 			}
+			r.removeNodeFromProcessing(workItem.nodeName)
```
Comment (deads2k) [inline pkg/controller/nodeipam/ipam/range_allocator.go:208]: In order to do this remove, you have to have a way to re-queue on a change.  The update handling between the two controllers appears to be different. See 

https://github.com/kubernetes/kubernetes/blob/4ae11dac2e2cc8594c528a679db5861c0926b514/pkg/controller/nodeipam/ipam/cloud_cidr_allocator.go#L117-L123

Comment (sjenning) [reply pkg/controller/nodeipam/ipam/range_allocator.go:208]: Are you making two different points here? The requeuing issue and the difference in update handling issue?

The update handling does seem to be different and no reason not to sync so it sets the NetworkUnavailable taint.

In terms of requeueing, I think you are right in that there is a `defer r.removeNodeFromProcessing(data.nodeName)` at the top of `updateCIDRsAllocation()` that needs to go.

**Comment by deads2k** [comment]
This appears to match now and lgtm.  I'd like to find someone from cloudcontroller manager to have a look too.  See if you can get them on slack.

**Comment by bowei** [comment]
/assign

is there a unit test to hit this code path?

**Comment by sjenning** [comment]
@bowei there isn't one, which is why this issue escaped in the first place.

I took some time to try to figure out how we could trick the code into running down this path by manipulating the `fakeNodeHandler` but didn't come up with anything yet.  Suggestions welcome!

**Comment by ehashman** [comment]
```
ERROR(linux/s390x): /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/controller/nodeipam/ipam/range_allocator.go:67:32: undeclared name: nodeProcessingInfo
... etc
```

typecheck CI is failing, @sjenning would you PTAL? The other failures look like flakes.

**Comment by sjenning** [comment]
Not sure what to do about the typecheck failure.  It compiles fine.  The `nodeProcessingInfo` type is private to the package.

**Comment by ehashman** [comment]
I don't think this can merge without fixing the typecheck...

/priority important-longterm
/triage accepted

**Comment by dims** [comment]
Is this PR still needed, please rebase if so (or we can close it?)

**Comment by dims** [comment]
This PR is older than 4 weeks and needs as rebase.

Please rebase the PR against latest master and reopen if still needed.

/close

**Pull request closed without merging**